Frontend/HsTypes.hs

```diff
@@ -142,6 +152,18 @@ instance (Symable a, PrettyPrint a) => PrettyPrint (HsAlt a) where
   ppr (HsAlt pat tm) = ppr pat <+> arrow <+> ppr tm
   needsParens _      = True
 
+instance (Symable a, PrettyPrint a) => PrettyPrint (HsPatGuard a) where
+  ppr (HsPatGuard p t) = ppr p <+> text "<-" <+> ppr t
+  needsParens _     = False
+
+instance (Symable a, PrettyPrint a) => PrettyPrint (HsGuarded a) where
+  ppr (HsGuarded [] t) = ppr t
+  ppr (HsGuarded gs t) = text "|"
+    <+> text (intercalate ", " (map (render . ppr) gs))
```

Comment:
Use only combinators if you can; don't go to String and then back to Doc. That is, do this instead:
```suggestion
    <+> fsep (punctuate comma (map ppr gs))
```